### PR TITLE
Implemented culture based authorization for content

### PR DESCRIPTION
**Pull request by Migaroez**
### Description
The Management API is lacking support to deny access to certain actions on resources based on the allowed cultures configured on User Groups. This PR aims to resolve this in the following way
- Expand the contentPermission Authorization design to take into account an optional Set of cultures on the ContentPermission resource. When set, all these cultures must be accessible to the user trough its configured groups
- Add this list to the auth request made on controller level for: document creation, document editing, document publishing and document unpublishing.
- Do something similar for Dictionary items create/update

### remarks
- It might be prudent to apply similar logic to
  - document deletion: if the user has access to all languages for which content data exists
  - document move/sort: if the user has access to at least one language for which content data exists
- The new Dictionary Permission design based on the content one is implemented with limited functionality and contains duplicated logic that is not abstracted away as we hope to change the way permissions are done in the near future. This system uses what is possible in the current design.

### Testing
- Add a new language
- Create a new invariant doctype with 1 invariant property (that is allowed on root)
- Test whether you can perform the following document actions in all languages
  - [ ] Create
  - [ ] Update
  - [ ] Publish
  - [ ] Unpublish
- Test whether you can perform the following dictionary actions in all languages
  - [ ] Create
  - [ ] Update
- Disable all language access for the admin group and only add one language to the allowed language list, do this directly in the DB as the current User Group post method is slightly broken when it comes to permissions. (Table name: umbracoUserGroup2Language)
- [ ] Test whether all previous tests still work for the allowed language
- [ ] Test whether all previous tests give a forbidden result for a language that is not in the allow list

### Automated test
- [ ] Because of the assumption (code suggested this) that only values (name/property) for cultures passed in the request (and thus update model) are actually updated on the content item, it seemed wise to add an integration test to verify this assumption.

### CODE REVIEW

File: src/Umbraco.Cms.Api.Management/Security/Authorization/Dictionary/DictionaryPermissionAuthorizer.cs [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+    {
+        _authorizationHelper = authorizationHelper;
+        _dictionaryPermissionService = dictionaryPermissionService;
+    }
+
+    public async Task<bool> IsAuthorizedForCultures(IPrincipal currentUser, ISet<string> culturesToCheck)
+    {
+        IUser user = _authorizationHelper.GetUmbracoUser(currentUser);
+        DictionaryAuthorizationStatus result = await _dictionaryPermissionService.AuthorizeCultureAccessAsync(user, culturesToCheck);
+        return result == DictionaryAuthorizationStatus.Success;
```
Comment (Zeegaan) [inline src/Umbraco.Cms.Api.Management/Security/Authorization/Dictionary/DictionaryPermissionAuthorizer.cs:23]:
```suggestion
        return result is DictionaryAuthorizationStatus.Success;
```

**Review by Zeegaan — CHANGES_REQUESTED** [review]
Found a few small things, but overall looks good 🙌

File: src/Umbraco.Core/Services/ContentPermissionService.cs [hunk 2]
```diff
@@ -130,6 +134,23 @@ public async Task<ContentAuthorizationStatus> AuthorizeBinAccessAsync(IUser user
...
+    public async Task<ContentAuthorizationStatus> AuthorizeCultureAccessAsync(IUser user, ISet<string> culturesToCheck)
+    {
+        if (user.Groups.Any(group => group.HasAccessToAllLanguages))
+        {
+            return ContentAuthorizationStatus.Success;
+        }
+
+        var allowedLanguageIsoCodes =
+            _languageService.GetIsoCodesByIds(user.Groups.SelectMany(g => g.AllowedLanguages)
+                .Distinct().ToArray());
```
[the remark this answers is not in the document]
Comment (Migaroez) [reply src/Umbraco.Core/Services/ContentPermissionService.cs:147]: There is a micro optimization for inlining functions whose result are only used as a parameter for another function. But readability is king so I yield 😉

File: src/Umbraco.Infrastructure/Persistence/Repositories/Implement/LanguageRepository.cs [hunk 3]
```diff
@@ -92,6 +92,38 @@ public LanguageRepository(IScopeAccessor scopeAccessor, AppCaches cache, ILogger
...
+        var isoCodes = new string[ids.Count];
+
+        if (ids.Any() == false)
+        {
+            return isoCodes;
+        }
+
+        EnsureCacheIsPopulated();
+
+        // yes, we want to lock _codeIdMap
```
[the remark this answers is not in the document]
Comment (Migaroez) [reply src/Umbraco.Infrastructure/Persistence/Repositories/Implement/LanguageRepository.cs]: I have updated all the comments regarding the lock and dictionary usage ;)

**Comment by Zeegaan** [comment]
Looks good, tests good!

**Pull request merged**